Doc updates (#1384)
intel/intel-device-plugins-for-kubernetes: cmd/gpu_plugin/README.md

| -enable-monitoring | - | disabled | Enable 'i915_monitoring' resource that provides access to all Intel GPU devices on the node |
| -resource-manager | - | disabled | Enable fractional resource management, [see also dependencies](#fractional-resources) |
| -shared-dev-num | int | 1 | Number of containers that can share the same GPU device |

> | -allocation-policy | string | none | 3 possible values: balanced, packed, none. For shared-dev-num > 1: balanced mode spreads workloads among GPU devices, packed mode fills one GPU fully before moving to next, and none selects first available device from kubelet. None mode is the default. Allocation policy does not have effect when resource manager is enabled. |

Reviewer: Your packed mode description improved things a lot.  I would still suggest:
```suggestion
| -allocation-policy | string | none | 3 possible values: balanced, packed, none. For shared-dev-num > 1: _balanced_ mode spreads workloads among GPU devices, _packed_ mode fills one GPU fully before moving to next, and _none_ selects first available device from kubelet. Default is _none_. Allocation policy does not have an effect when resource manager is enabled. |
```

The plugin also accepts a number of other arguments (common to all plugins) related to logging.
Please use the -h option to see the complete list of logging related options.

> ## Use Cases for Different Modes

Reviewer: Because mode is based on workloads, and not the other way round, I would suggest:
```suggestion
## Operation modes for different workload types
```

> Intel GPU-plugin supports a few different operation modes. Depending on the workloads the cluster is running, some modes make less sense than others. Below is a table that explains the differences between the modes and suggests workload types for each mode. The mode selection requires pre-though as it is cluster wide.

Reviewer: Not sure whether this is better...
```suggestion
Intel GPU-plugin supports a few different operation modes. Depending on the workloads the cluster is running, some modes make less sense than others. Below is a table that explains the differences between the modes and suggests workload types for each mode. Mode selection applies to the whole GPU plugin deployment, so it is a cluster wide decision.
```

> | Mode | Sharing | Intended workloads | Time critical |

Reviewer: ```suggestion
| Mode | Sharing | Intended workloads | Suitable for time critical workloads |
```

|:---- |:-------- |:------- |:------- |
| shared-dev-num == 1 | No, 1 container per GPU | Workloads using all GPU capacity, e.g. AI training | Yes |

> | shared-dev-num > 1 | Yes, >1 containers per GPU | (Batch) workloads using only part of GPU resources, e.g. inference, media transcode/analytics | No |

Reviewer: Batch workloads often use whole GPU (as they run as fast as they can) _unless_ parallel workloads can use completely different GPU pipeline (e.g. *fixed* i.e. low-power media pipeline vs. compute), or they are CPU bound:
```suggestion
| shared-dev-num > 1 | Yes, >1 containers per GPU | (Batch) workloads using only part of GPU resources, e.g. inference, media transcode/analytics / CPU bound GPU workloads | No |
```

(Whereas real-time workloads often have only partial GPU usage, but are timing critical i.e. one really needs GAS.)

PS. I'm not sure whether this mode is something that should be "recommended" to end users, as WLs for which it is useful are so specific.

> | shared-dev-num > 1 | Yes, >1 containers per GPU | (Batch) workloads using only part of GPU resources, e.g. inference, media transcode/analytics / CPU bound GPU workloads | No |

Reviewer: Hopefully last fine-tuning comments (it's a bit embarrasing to be fixing my own suggestions)...

Two `/` close together did not look good:
```suggestion
| shared-dev-num > 1 | Yes, >1 containers per GPU | (Batch) workloads using only part of GPU resources, e.g. inference, media transcode/analytics, or CPU bound GPU workloads | No |
```

Note: I'm assuming the reason why your initially listed workloads do not use whole GPU is not due to them be being CPU bound, but due to their input data rate being limited (coming e.g. from camera at certain FPS).

> | shared-dev-num > 1 && resource-management | Yes and no, 1>= containers per GPU | Any. For best results, all workloads should declare their expected GPU resource usage (memory, millicores). Requires [GAS](https://github.com/intel/platform-aware-scheduling/tree/master/gpu-aware-scheduling). See also [fractional use](#fractional-resources-details). | Depends on the requested GPU resources |

Reviewer: It could be more explicit:
```suggestion
| shared-dev-num > 1 && resource-management | Yes and no, 1>= containers per GPU | Any. For best results, all workloads should declare their expected GPU resource usage (memory, millicores). Requires [GAS](https://github.com/intel/platform-aware-scheduling/tree/master/gpu-aware-scheduling). See also [fractional use](#fractional-resources-details). | 1000 millicores = exclusive GPU usage |
```

> | shared-dev-num > 1 && resource-management | Yes and no, 1>= containers per GPU | Any. For best results, all workloads should declare their expected GPU resource usage (memory, millicores). Requires [GAS](https://github.com/intel/platform-aware-scheduling/tree/master/gpu-aware-scheduling). See also [fractional use](#fractional-resources-details). | 1000 millicores = exclusive GPU usage |

Reviewer: Removing "." and adding "Yes" for consistency with other entries:
```suggestion
| shared-dev-num > 1 && resource-management | Yes and no, 1>= containers per GPU | Any. For best results, all workloads should declare their expected GPU resource usage (memory, millicores). Requires [GAS](https://github.com/intel/platform-aware-scheduling/tree/master/gpu-aware-scheduling). See also [fractional use](#fractional-resources-details) | Yes. 1000 millicores = exclusive GPU usage |
```

| -enable-monitoring | - | disabled | Enable 'i915_monitoring' resource that provides access to all Intel GPU devices on the node |
| -resource-manager | - | disabled | Enable fractional resource management, [see also dependencies](#fractional-resources) |
| -shared-dev-num | int | 1 | Number of containers that can share the same GPU device |
| -allocation-policy | string | none | 3 possible values: balanced, packed, none. For shared-dev-num > 1: _balanced_ mode spreads workloads among GPU devices, _packed_ mode fills one GPU fully before moving to next, and _none_ selects first available device from kubelet. Default is _none_. Allocation policy does not have an effect when resource manager is enabled. |

The plugin also accepts a number of other arguments (common to all plugins) related to logging.
Please use the -h option to see the complete list of logging related options.

## Operation modes for different workload types

> Intel GPU-plugin supports a few different operation modes. Depending on the workloads the cluster is running, some modes make less sense than others. Below is a table that explains the differences between the modes and suggests workload types for each mode. Mode selection applies to the whole GPU plugin deployment, so it is a cluster wide decision.

Reviewer 2: I would use "more" rather than "less", even though in your shell "less" may be a better alternative.

| Mode | Sharing | Intended workloads | Suitable for time critical workloads |
|:---- |:-------- |:------- |:------- |
| shared-dev-num == 1 | No, 1 container per GPU | Workloads using all GPU capacity, e.g. AI training | Yes |
| shared-dev-num > 1 | Yes, >1 containers per GPU | (Batch) workloads using only part of GPU resources, e.g. inference, media transcode/analytics, or CPU bound GPU workloads | No |

> | shared-dev-num > 1 && resource-management | Yes and no, 1>= containers per GPU | Any. For best results, all workloads should declare their expected GPU resource usage (memory, millicores). Requires [GAS](https://github.com/intel/platform-aware-scheduling/tree/master/gpu-aware-scheduling). See also [fractional use](#fractional-resources-details) | Yes. 1000 millicores = exclusive GPU usage |

Reviewer 2: Usage of the extended resources doesn't guarantee exclusive usage, unless all workloads consume the same resource. A rogue Pod with 0 millicores will fit into a GPU which has 1000 already used.

Author: I added a footnote.

> PS. IMHO containers that request GPU(s), but do not declare both GPU resources usage for GAS, should be treated as requesting 1000* millicores. I mean, if GPU container memory usage is not specified, GAS should not assume that it fits to same GPU with the other GPU containers. That's just not reliable.

I would have to agree.

Intel GPU-plugin supports a few different operation modes. Depending on the workloads the cluster is running, some modes make more sense than others. Below is a table that explains the differences between the modes and suggests workload types for each mode. Mode selection applies to the whole GPU plugin deployment, so it is a cluster wide decision.

| Mode | Sharing | Intended workloads | Suitable for time critical workloads |
|:---- |:-------- |:------- |:------- |
| shared-dev-num == 1 | No, 1 container per GPU | Workloads using all GPU capacity, e.g. AI training | Yes |
| shared-dev-num > 1 | Yes, >1 containers per GPU | (Batch) workloads using only part of GPU resources, e.g. inference, media transcode/analytics, or CPU bound GPU workloads | No |
| shared-dev-num > 1 && resource-management | Yes and no, 1>= containers per GPU | Any. For best results, all workloads should declare their expected GPU resource usage (memory, millicores). Requires [GAS](https://github.com/intel/platform-aware-scheduling/tree/master/gpu-aware-scheduling). See also [fractional use](#fractional-resources-details) | Yes. 1000 millicores = exclusive GPU usage. See note below. |

> > **Note**: With resource management, there are corner cases where an exlusive GPU can be used by other Pods. For example, a privileged Pod or a Pod requesting 0 millicores might get access to a fully utilized GPU.

Reviewer: Most common case where this is goes wrong is containers asking for GPU but not declaring how much resources they use, so I would state this simply as:
```suggestion
> **Note**: Exclusive GPU usage with >=1000 millicores requires that also *all other GPU containers* specify (non-zero) millicores resource usage.
```
This being about resource management should be obvious.

      Warning  FailedScheduling  <unknown>  default-scheduler  0/1 nodes are available: 1 Insufficient gpu.intel.com/i915.
    ```

## Use Cases for Different Modes

> Intel GPU-plugin supports a few different operation modes. Depending on the workloads the cluster is running, some modes make less sense than others. Below is a table that explains the differences between the modes and suggests workload types for each mode. The mode selection requires pre-though as it is cluster wide.

Reviewer: Fair enough.  This can be resolved, I added few additional text fine-tuning suggestions.

> | Mode | Sharing | Workload examples | Time critical |

Reviewer: "examples" may not be best term here, as column describes *all* workloads this mode is intended for.  Maybe "Intended workloads"?

I assume "Time critical" to mean "Whether workloads is latency sensitive".

|:---- |:-------- |:------- |:------- |
| shared-dev-num == 1 | No, 1 container per GPU | AI/ML training or an application that can fully utilize a GPU e.g. pack multiple workloads. | Yes |
| shared-dev-num > 1 | Yes, >1 containers per GPU | Inference, media transcode, media analytics. Workloads that only require part of the GPU. See also [allocation profiles](#modes-and-configuration-options). | No |

> | shared-dev-num > 1 && resource-management | Yes and no, 1>= containers per GPU | All workloads, but requires GPU resource allocations. GPUs can be dedicated and shared based on the GPU resources like memory and millicores. Requires [GAS](https://github.com/intel/platform-aware-scheduling/tree/master/gpu-aware-scheduling). See also [fractional use](#fractional-resources-details). | Depends on the Pod spec |

Reviewer: First state the intended workload type, and then give examples of them, for example:
* "AI/ML training or an application that can fully utilize a GPU e.g. pack multiple workloads" -> "Workloads using all GPU capacity, e.g. AI training"
* "Inference, media transcode, media analytics. Workloads that only require part of the GPU" -> "(Batch) workloads using only part of GPU resources, e.g. inference, media transcode/analytics"
* "All workloads, but requires GPU resource allocations. GPUs can be dedicated and shared based on the GPU resources like memory and millicores" -> "Any.  For best results, *all* workloads should declare their expected GPU resource usage (memory, millicores)"

What means "Time critical" depending pod spec??

Author: Thanks.
I rewrote "depending on pod spec" as "Depends on the requested GPU resources".
